RecoTracker/TkTrackingRegions/plugins/L1MuonSeededTrackingRegionsProducer.h

```diff
@@ -0,0 +1,411 @@
+    descRegion.add<edm::InputTag>("measurementTrackerName", edm::InputTag(""));
+    descRegion.add<bool>("searchOpt", false);
+    desc.add<edm::ParameterSetDescription>("RegionPSet", descRegion);
+
+    // MuonServiceProxy for the propagation
+    edm::ParameterSetDescription psd0;
+    psd0.addUntracked<std::vector<std::string>>("Propagators", {"SteppingHelixPropagatorAny"});
+    psd0.add<bool>("RPCLayers", false);
+    psd0.addUntracked<bool>("UseMuonNavigation", false);
+    desc.add<edm::ParameterSetDescription>("ServiceParameters", psd0);
+
+    descriptions.add("hltIterL3MuonPixelTracksTrackingRegions", desc);
+  }
+
+  std::vector<std::unique_ptr<TrackingRegion>> regions(const edm::Event& iEvent,
+                                                       const edm::EventSetup& iSetup) const override {
+    std::vector<std::unique_ptr<TrackingRegion>> result;
+
+    // pick up the candidate objects of interest
+    edm::Handle<l1t::MuonBxCollection> muColl;
+    iEvent.getByToken(token_input, muColl);
+    if (muColl->size() == 0)
+      return result;
+
+    // always need the beam spot (as a fall back strategy for vertex modes)
+    edm::Handle<reco::BeamSpot> bs;
+    iEvent.getByToken(token_beamSpot, bs);
+    if (!bs.isValid())
+      return result;
+
+    // this is a default origin for all modes
+    GlobalPoint default_origin(bs->x0(), bs->y0(), bs->z0());
+
+    // vector of origin & halfLength pairs:
+    std::vector<std::pair<GlobalPoint, float>> origins;
+
+    // fill the origins and halfLengths depending on the mode
+    if (m_mode == BEAM_SPOT_FIXED || m_mode == BEAM_SPOT_SIGMA) {
+      origins.push_back(std::make_pair(
+          default_origin, (m_mode == BEAM_SPOT_FIXED) ? m_zErrorBeamSpot : m_nSigmaZBeamSpot * bs->sigmaZ()));
+    } else if (m_mode == VERTICES_FIXED || m_mode == VERTICES_SIGMA) {
+      edm::Handle<reco::VertexCollection> vertices;
+      iEvent.getByToken(token_vertex, vertices);
+      int n_vert = 0;
+      for (reco::VertexCollection::const_iterator v = vertices->begin();
+           v != vertices->end() && n_vert < m_maxNVertices;
+           ++v) {
+        if (v->isFake() || !v->isValid())
+          continue;
+
+        origins.push_back(std::make_pair(GlobalPoint(v->x(), v->y(), v->z()),
+                                         (m_mode == VERTICES_FIXED) ? m_zErrorVetex : m_nSigmaZVertex * v->zError()));
+        ++n_vert;
+      }
+      // no-vertex fall-back case:
+      if (origins.empty()) {
+        origins.push_back(std::make_pair(
+            default_origin, (m_nSigmaZBeamSpot > 0.) ? m_nSigmaZBeamSpot * bs->z0Error() : m_zErrorBeamSpot));
+      }
+    }
+
+    const MeasurementTrackerEvent* measurementTracker = nullptr;
+    if (!token_measurementTracker.isUninitialized()) {
+      edm::Handle<MeasurementTrackerEvent> hmte;
+      iEvent.getByToken(token_measurementTracker, hmte);
+      measurementTracker = hmte.product();
```

Comment:
```suggestion
      measurementTracker = &iEvent.get(token_measurementTracker);
```
I think this whole section could be simplified.

Reply:
done